### DISCUSSION TITLE
Combining null-conditional member access and null-coalescing with `ref struct`

### DISCUSSION THREAD

**Discussion by PetSerAl** [dotnet/csharplang#7003] [Language Ideas]
```csharp
struct NormalStruct { }
ref struct RefStruct { }
class Class {
    public NormalStruct NormalStruct => default;
    public RefStruct RefStruct => default;
}
class Test {
    NormalStruct M11(Class? c) => c?.NormalStruct ?? default;
    NormalStruct M12(Class? c) => c is not null ? c.NormalStruct : default;
    RefStruct M21(Class? c) => c?.RefStruct ?? default;
    RefStruct M22(Class? c) => c is not null ? c.RefStruct : default;
}
```
C# compiler reject `M21` method because it think, that subexpression `c?.RefStruct` have type `Nullable<RefStruct>`, which is not allowed. Although expression as whole never cause any instance of `Nullable<RefStruct>` to be created, and semantically equivalent method `M22` compiled without any error.
Should compiler allow this code pattern for `ref struct`?

**Comment by 333fred** [comment]
> Although expression as whole never cause any instance of Nullable<RefStruct> to be created, and semantically equivalent method M22 compiled without any error.

Regardless of any compiler optimizations, this is not how the _language_ defines this codepattern. The type of `c?.NormalStruct` is `Nullable<NormalStruct>`, and roslyn will happily tell you this if you inspect the code with the public API. While you could update the spec to allow this, it would be a complicated change, both in the specification and in the actual compiler itself.
